direct/core/src/main/java/cz/o2/proxima/direct/time/PartitionedWatermarkEstimator.java

```diff
@@ -0,0 +1,59 @@
+/**
+ * Copyright 2017-2020 O2 Czech Republic, a.s.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package cz.o2.proxima.direct.time;
+
+import cz.o2.proxima.storage.StreamElement;
+import cz.o2.proxima.time.WatermarkEstimator;
+import cz.o2.proxima.time.WatermarkSupplier;
+import java.util.Collections;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Watermark estimator wrapper for partitioned sources. Estimates watermark as a minimum across all
+ * partition's watermarks. Uses processing time when none partitions are assigned. Update and idle
+ * calls are delegated to estimator for a given partition.
+ */
+public class PartitionedWatermarkEstimator implements WatermarkSupplier {
+  private static final long serialVersionUID = 1L;
+  private final ConcurrentHashMap<Integer, WatermarkEstimator> estimators;
+
+  public PartitionedWatermarkEstimator(Map<Integer, WatermarkEstimator> partitionsEstimators) {
+    estimators = new ConcurrentHashMap<>(partitionsEstimators);
+  }
+
+  public static PartitionedWatermarkEstimator empty() {
+    return new PartitionedWatermarkEstimator(Collections.emptyMap());
```

Comment:
This looks dangerous. I'd remove this, the `PartitionedWatermarkEstimator.empty()` seems to be `ProcessingTimeWatermarkEstimator` and we should not mix them.

Reply:
Agree, but this is just backward compatible solution. KafkaLogReader and it unit tests requires that  AtomicReference is initialized and watermark supplier returns processing time when no partitions are assigned.

Reply:
That sounds like the test might be inadequate. We might consider changing the tests.

Reply:
Not test but KafkaLogReader. 

This snippet is called earlier
```
 if (poll.isEmpty()) {
                consumer.onIdle(watermarkEstimator.get());
 }
```
than

```
increaseWatermarkOnEmptyPolls(emptyPollCount, watermarkEstimator);
```

It causes that consumer gets information about idle earlier than idle-policy. Just reordering doesn't help because  emptyPollCount logic.

Reply:
Will be addressed in O2-Czech-Republic#155